### ISSUE TITLE
Unexpected Behavior with Optional Chaining and Array Access

### ISSUE THREAD

**Issue by MarcusTF** [microsoft/TypeScript#57025]
### 🔎 Search Terms

"optional chaining", "array access", "undefined index", "strictNullChecks", "TypeScript array optional chaining"

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about optional chaining, array index access, etc.

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/C4TwDgpgBAKhDOwoF4oG8BQUoEMBOeOIA-AFxQCCBRAPJttgG74CWOARgDYRlSJ4sAdgHMsUAL4A+DOIwYAxgHtBiKMATBycVanq5qIcgG0AugBoZcgPRWoAUQAekeUhjgIAZXkCwSYIqgAd3xBKAALCDxoHHg1DQA6fEISeKMABhMoJQBXTgATKHZobME8iAAzIQg8hWVVKPhcpFR1RESDYlSM+OYBDm4AblqVRW54zkVhAAoGpoBKIaA

### 💻 Code

```ts
type Test = {
  array?: Array<{
    variable?: string
  }>
}

const test: Test = {
  array: [],
}

// Expect TypeScript to warn here as test.array?.[0] could be undefined
const result = test.array?.[0].variable;
console.log(result);
```

### 🙁 Actual behavior

TypeScript does not warn that `test.array?.[0]` could be undefined, leading to potential runtime errors when accessing `.variable`.

### 🙂 Expected behavior

TypeScript should consistently issue warnings about potential `undefined` values when accessing properties on objects that are obtained through optional chaining, regardless of whether the property access is done directly in the chained expression or in subsequent steps. The discrepancy observed in the handling of `test.array?.[0].variable` versus the broken-down approach (`index0.variable`) highlights an inconsistency in TypeScript's type safety checks. This inconsistency could lead to confusion and potential bugs, especially in complex codebases where developers rely on TypeScript for identifying such issues.

### Additional information about the issue

As this is my first issue opened in the TypeScript repository, I apologize in advance if this is a known issue, a design limitation, or an intentional choice. I have attempted to search for more information on this topic but was unable to find conclusive references that could clarify whether the behavior I've encountered is expected or known.

This issue raises concerns about the consistency of TypeScript's type safety checks, particularly under `strictNullChecks`. In similar contexts, TypeScript typically warns about potential `undefined` values, which is a key aspect of its type safety enforcement. However, this expected behavior seems to be inconsistently applied in the scenario involving optional chaining with array access.

### Deeply Nested Object Comparison:
Consider the following TypeScript code with a deeply nested object structure:

```typescript
type DeepNestedType = {
  level1?: {
    level2?: {
      level3?: {
        property: string;
      }
    }
  }
}

const deepNestedObject: DeepNestedType = {};

// TypeScript should warn here as any of the levels could be undefined
const notOkay1 = deepNestedObject.level1.level2.level3.property; 
const notOkay2 = deepNestedObject?.level1.level2.level3.property;
const notOkay3 = deepNestedObject.level1?.level2.level3.property;
const notOkay4 = deepNestedObject.level1.level2?.level3.property;
const notOkay5 = deepNestedObject.level1?.level2?.level3.property;

const okay = deepNestedObject.level1?.level2?.level3?.property;
```

In these examples, TypeScript typically provides warnings for accessing properties on potentially `undefined` values, as seen in `notOkay1` through `notOkay5`. This aligns with TypeScript's behavior under `strictNullChecks`, where direct property access without optional chaining or proper checks is flagged.

#### Contrasting Behavior with Array Example:
Conversely, in the array example (`test.array?.[0].variable`), TypeScript does not issue a similar warning when a property is accessed on a potentially `undefined` array element. This discrepancy in handling deeply nested objects and array elements leads to inconsistency in TypeScript's type safety checks.

### Behavior with Separated Optional Chaining Steps:
An additional observation is that when optional chaining is separated into distinct steps, TypeScript's type inference behaves as expected:

```typescript
type Test = {
  array?: Array<{
    variable?: string
  }>
}

const test: Test = {
  array: [],
}

const index0 = test.array?.[0]; // TypeScript correctly infers this as undefined | { variable?: string }
const variable = index0.variable; // TypeScript issues a warning here, as expected
```

In this scenario, TypeScript correctly infers the type of `index0` and warns about the potential `undefined` value when accessing `.variable`. This is in contrast to the original example, where such a warning is not provided.

These observations highlight an inconsistency in TypeScript's handling of optional chaining and undefined checks. Such inconsistency can lead to confusion and potential bugs in codebases, particularly in complex scenarios where developers rely on TypeScript for enhanced type safety. A consistent approach to handling potential undefined values in all scenarios involving optional chaining would greatly improve the reliability and predictability of TypeScript's type safety checks.

**Comment by fatcerberus** [comment]
> An additional observation is that when optional chaining is separated into distinct steps, TypeScript's type inference behaves as expected

That's expected.  These are not equivalent:
```ts
let x = a?.b.c;
```
vs.
```ts
let y = a?;
let x = x.b.c;
```
because optional chaining short circuits.

As for this part:
> Expect TypeScript to warn here as test.array?.[0] could be undefined

Array accesses are assumed to be in-bounds.  See compiler option `noUncheckedIndexedAccess`.

**Comment by MarcusTF** [comment]
Thank you so much for the fast response and clarification ❤️
